### PR TITLE
Implement native staking

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
- Implements [HIP-406](https://hips.hedera.com/hip/hip-406).

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/ServicesState.java [hunk 1]
```diff
@@ -174,9 +178,12 @@ public int getVersion() {
 
 	@Override
 	public int getMinimumChildCount(int version) {
-		if (version >= MINIMUM_SUPPORTED_VERSION && version <= CURRENT_VERSION) {
+		if (version >= MINIMUM_SUPPORTED_VERSION && version < CURRENT_VERSION) {
 			return NUM_POST_0210_CHILDREN;
-		} else {
+		} else if (version == CURRENT_VERSION) {
+			return NUM_POST_0270_CHILDREN;
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/ServicesState.java:184]: This should be renamed to `NUM_0270_CHILDREN`

**Review by Neeharika-Sompalli — COMMENTED** [review]
Just thought, should the fees for cryptoCreate , contractCreate , cryptoUpdate , contractUpdate change in Rbs and BPT by a long value ?

File: hedera-node/src/main/java/com/hedera/services/ServicesState.java (hunk 1 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/ServicesState.java]: Fixed, also bumped minimum supported version to 0.25.x

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java [hunk 2]
```diff
@@ -80,7 +82,8 @@ public class MerkleNetworkContext extends AbstractMerkleLeaf {
 	static final int RELEASE_0200_VERSION = 6;
 	static final int RELEASE_0240_VERSION = 7;
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java:83]: Do we need migration for these two versions here and `getMinimumSupportdVersion` should change?

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleStakingInfo.java [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+import java.util.Objects;
+
+import static com.hedera.services.ServicesState.EMPTY_HASH;
+import static com.hedera.services.legacy.proto.utils.CommonUtils.noThrowSha384HashOf;
+import static com.hedera.services.state.merkle.internals.ByteUtils.getHashBytes;
+
+public class MerkleStakingInfo extends AbstractMerkleLeaf implements Keyed<EntityNum> {
+	private static final Logger log = LogManager.getLogger(MerkleStakingInfo.class);
+	// get this max history from the props.
+	static final int MAX_REWARD_HISTORY = 366;
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleStakingInfo.java:52]: This value should be set in line 79 as `numRewardablePeriods + 1` right ?

**Review by Neeharika-Sompalli — APPROVED** [review]
Other than the above two comments , rest LGTM !

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleStakingInfo.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleStakingInfo.java]: Eliminated use of this constant. (Was only in `deserialize()` tho.)

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java (hunk 2 above)
Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleNetworkContext.java]: Bumped min supported to 0.24.x

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM !

**Pull request merged**